src/Raven.Client/Http/FailoverNodeSelector.cs

```diff
@@ -76,17 +71,35 @@ public void RestoreNodeIndex(int nodeIndex)
             }
         }
 
+        public virtual INodeSelector CloneForNewSession()
+        {
+            //prevent race condition where _topology has already changed but _currentNodeIndex has not yet changed
+            lock (_cloneForNewSessionSync)
```

Comment:
No, this is _bad_, what you ensured will happen is that you are going to _force_ locking on pretty much every request.
That is a really bad idea.